### PR TITLE
Fix CSE across unrelated mutation regions

### LINKED ISSUE #174472: [Inductor][CSE] `split_with_sizes` not deduplicated on B200 with FP8 quantization
### Describe the bug

**Summary:**
When compiling models with FP8 quantization on NVIDIA B200 GPUs, Inductor's CSE pass fails to deduplicate identical `split_with_sizes` calls. The graph contains 3 separate split nodes with 1 user each, instead of 1 split node with 3 users.

**Context:**
- Affects: CSE pass in Inductor
- GPU: NVIDIA B200 (Blackwell)
- Dtype: FP8 (via vLLM's quantization path)
- Impact: Breaks downstream pattern matching that expects deduplicated nodes

### Minimal Reproducible Example

```python
import torch

# Simulates the pattern seen in vLLM's QK norm + RoPE fusion
# On B200+FP8, the three split_with_sizes calls are NOT deduplicated by CSE

def model(qkv):
    # These should be CSE'd into a single split node
    split1 = torch.ops.aten.split_with_sizes.default(qkv, [512, 128, 128], dim=-1)
    split2 = torch.ops.aten.split_with_sizes.default(qkv, [512, 128, 128], dim=-1)
    split3 = torch.ops.aten.split_with_sizes.default(qkv, [512, 128, 128], dim=-1)
    
    q = split1[0]  # First user
    k = split2[1]  # Second user  
    v = split3[2]  # Third user
    return q, k, v

compiled = torch.compile(model, backend="inductor")
qkv = torch.randn(1024, 768, dtype=torch.bfloat16, device="cuda")
compiled(qkv)
```

*Note: Actual repro requires B200 hardware with FP8 model. The issue was observed in vLLM with `Qwen/Qwen3-30B-A3B-FP8`.*

### Expected Behavior

CSE should recognize the three `split_with_sizes` calls are identical (same input tensor, same split sizes, same dim) and deduplicate them into a single node with 3 `getitem` users.

### Actual Behavior

Three separate `split_with_sizes` nodes remain in the FX graph:

```
split_with_sizes_1 = torch.ops.aten.split_with_sizes.default(qkv, [512, 128, 128], -1)
getitem_0 = split_with_sizes_1[0]  # q

split_with_sizes_2 = torch.ops.aten.split_with_sizes.default(qkv, [512, 128, 128], -1)
getitem_1 = split_with_sizes_2[1]  # k

split_with_sizes_3 = torch.ops.aten.split_with_sizes.default(qkv, [512, 128, 128], -1)
getitem_2 = split_with_sizes_3[2]  # v
```

This breaks pattern matchers that expect `_users=3` on a single split node.

### Ablation Testing

- [x] Works correctly on: H100 with BF16/FP16
- [ ] Fails on: B200 with FP8 quantization
- [ ] Likely related to FP8 tensor handling or Blackwell-specific IR generation

### Versions

```
PyTorch: 2.9.1+cu129
GPU: NVIDIA B200
CUDA: 12.9
Python: 3.12
```

### Additional Context

**Workaround:** vLLM implemented a pre-pass to manually coalesce duplicate splits before pattern matching: https://github.com/vllm-project/vllm/pull/33967

**Related vLLM issue:** https://github.com/vllm-project/vllm/issues/33295

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @jansel @eellison @zou3519 @BoyuanFeng @angelayi

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #183969

Allow fx_graph_cse to reuse duplicate pure aten ops across mutation region boundaries when all intervening mutable aten writes are known not to alias the candidate inputs or the previous duplicate result. Add regression coverage for unrelated writes, input aliases, unknown mutations, and result mutation safety.

Fixes #174472

Generated by my agent

cc @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by yushangdi** [comment]
looks like there is already a PR for the same issue at https://github.com/pytorch/pytorch/pull/175836

**Comment by jansel** [comment]
My agent says #175836 is an earlier unmerged/stale attempt at the same root cause, but #183969 is not redundant as written. The older patch only protects the candidate input storage, so it can incorrectly CSE if the previous duplicate result was mutated in between (for example `cos_1 = cos(x); copy_(cos_1, src); cos_2 = cos(x)`). This PR also protects the previous duplicate output, checks only schema-written mutation arguments, fails closed for unknown/non-aten mutation barriers, and adds tests for those cases.

File: torch/_functorch/compile_utils.py [hunk 1]
```diff
@@ -47,6 +47,149 @@ def get_aten_target(node: fx.Node) -> OpOverloadPacket | Callable[..., Any] | st
...
+    when the mutation is completely unrelated to the op being considered for
+    CSE.  This helper performs a finer-grained check for pure aten ops.
+    """
+    if node.target != prev_node.target:
+        return False
+    if not isinstance(node.target, torch._ops.OpOverload):
+        return False
+    if node.target.namespace != "aten" or node.target._schema.is_mutable:
+        return False
+    if prev_node.meta["mutation_region_id"] > node.meta["mutation_region_id"]:
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py:208]: [cleanup] This strict-`>` guard is unreachable. `old_node_for_hash[hash_val]` is only written inside this same loop (line 322), so `prev_node` is always a node we already processed — strictly earlier in graph iteration order than `node`. `compute_mutation_region_ids` assigns ids monotonically non-decreasing in graph order, so `prev_node.meta['mutation_region_id'] <= node.meta['mutation_region_id']` is invariant here. Either drop the check or replace it with an assertion to make the invariant explicit.

File: torch/_functorch/compile_utils.py [hunk 2]
```diff
@@ -172,6 +305,11 @@ def substitute(
                 if same_mutation_regions(n, duplicate_n_prev):
                     env[n] = duplicate_n_prev
                     continue
+                elif _can_cse_across_mutation_regions(
+                    n, old_node_for_hash[hash_val], is_mutation_op
+                ):
+                    env[n] = duplicate_n_prev
+                    continue
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py:455]: [question] On the cross-region success path you (correctly) avoid updating `hash_env` / `old_node_for_hash`, but that means future candidates with this hash will keep re-walking from the *original* earliest occurrence. Worst case: a long chain `cos(x); unrelated_mutation; cos(x); unrelated_mutation; cos(x); ...` — each new `cos(x)` walks all the way back to the first one, so total cross-region work is O(K · N) ≈ O(N²) for K duplicates. Each walked step also re-runs `_get_mutated_argument_values` + `_collect_storage_refs_from_graph_values` per intervening mutation. Worth a thought on whether to advance `old_node_for_hash[hash_val] = n` on success too (the storages-to-protect are identical for any future same-hash candidate, so the already-cleared prefix can't become unsafe). Could matter on the kinds of large graphs that motivated this PR.

File: torch/_functorch/compile_utils.py [hunk 3]
```diff
@@ -47,6 +47,149 @@ def get_aten_target(node: fx.Node) -> OpOverloadPacket | Callable[..., Any] | st
 ]
 
 
+def _storage_refs_from_value(value: Any) -> tuple[set[StorageWeakRef], bool]:
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py]: nit:

```suggestion
def _storage_refs_from_value(value: object) -> tuple[set[StorageWeakRef], bool]:
```

File: torch/_functorch/compile_utils.py [hunk 4]
```diff
@@ -47,6 +47,149 @@ def get_aten_target(node: fx.Node) -> OpOverloadPacket | Callable[..., Any] | st
...
+    storage_refs: set[StorageWeakRef] = set()
+    for leaf in tree_flatten(value)[0]:
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py]: nit:

```suggestion
    for leaf in tree_iter(value):
```

File: torch/_functorch/compile_utils.py [hunk 5]
```diff
@@ -47,6 +47,149 @@ def get_aten_target(node: fx.Node) -> OpOverloadPacket | Callable[..., Any] | st
...
+        if not isinstance(leaf, torch.Tensor):
+            continue
+        try:
+            storage_refs.add(StorageWeakRef(leaf.untyped_storage()))
+        except NotImplementedError:
+            return set(), True
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py]: nit: tighter try/except:

```suggestion
        try:
            storage = leaf.untyped_storage()
        except NotImplementedError:
            return set(), True
        storage_refs.add(StorageWeakRef(storage))
```

File: torch/_functorch/compile_utils.py (hunk 3 above)
Comment (aorenste) [inline torch/_functorch/compile_utils.py]: Might be slightly cleaner to retrn `set[...] | None` rather than a tuple.

File: torch/_functorch/compile_utils.py [hunk 6]
```diff
@@ -47,6 +47,149 @@ def get_aten_target(node: fx.Node) -> OpOverloadPacket | Callable[..., Any] | st
...
+    except NotImplementedError:
+        return False
+    return True
+
+
+def _collect_storage_refs_from_graph_values(
+    values: Any,
+) -> set[StorageWeakRef] | None:
+    storage_refs: set[StorageWeakRef] = set()
+    for value in tree_flatten(values)[0]:
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py]: nit:
```suggestion
    for value in tree_iter(values):
```

File: torch/_functorch/compile_utils.py [hunk 7]
```diff
@@ -47,6 +47,149 @@ def get_aten_target(node: fx.Node) -> OpOverloadPacket | Callable[..., Any] | st
...
+    if (
+        not isinstance(node.target, torch._ops.OpOverload)
+        or node.target.namespace != "aten"
+        or not node.target._schema.is_mutable
+    ):
+        return None
+
+    mutated_values: list[Any] = []
+    positional_idx = 0
+    missing = object()
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py]: nit: make this a global?

File: torch/_functorch/compile_utils.py [hunk 8]
```diff
@@ -47,6 +47,149 @@ def get_aten_target(node: fx.Node) -> OpOverloadPacket | Callable[..., Any] | st
...
+            value_storage_refs, has_unavailable_storage = _storage_refs_from_value(
+                value
+            )
+            if has_unavailable_storage:
+                return None
+            storage_refs.update(value_storage_refs)
+    return storage_refs
+
+
+def _get_mutated_argument_values(node: fx.Node) -> Any | None:
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py]:
```suggestion
def _get_mutated_argument_values(node: fx.Node) -> tuple[Any] | None:
```

File: torch/_functorch/compile_utils.py [hunk 9]
```diff
@@ -47,6 +47,149 @@ def get_aten_target(node: fx.Node) -> OpOverloadPacket | Callable[..., Any] | st
...
+        return False
+
+    storage_refs_to_protect = input_storage_refs | prev_output_storage_refs
+    if not storage_refs_to_protect:
+        return True
+
+    cur = prev_node.next
+    for _ in range(len(prev_node.graph.nodes)):
+        if cur is node:
+            return True
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py]: This is a weirdly defensive way to say
```
while cur is not node:
```

File: torch/_functorch/compile_utils.py (hunk 3 above)
Comment (aorenste) [inline torch/_functorch/compile_utils.py]: I found it confusing we were using StorageWeakRef here because we don't actually care about weakrefs. Consider clarifying it by adding a `StorageKey = StorageWeakRef` alias (or newtype) and using that to document the behavior (but use it i the old functions too). Or at least a comment to indicate why we're using StorageWeakRef.

File: torch/_functorch/compile_utils.py [hunk 10]
```diff
@@ -47,6 +47,149 @@ def get_aten_target(node: fx.Node) -> OpOverloadPacket | Callable[..., Any] | st
...
+        if is_mutation_op_fn(cur):
+            mutated_values = _get_mutated_argument_values(cur)
+            if mutated_values is None:
+                return False
+            mutated_storage_refs = _collect_storage_refs_from_graph_values(
+                mutated_values
+            )
+            if mutated_storage_refs is None or not mutated_storage_refs:
+                return False
+            if mutated_storage_refs & storage_refs_to_protect:
```
Comment (aorenste) [inline torch/_functorch/compile_utils.py]:
```suggestion
            if not mutated_storage_refs.isdisjoint(storage_refs_to_protect):
```

**Review by aorenste — CHANGES_REQUESTED** [review]

File: torch/_functorch/compile_utils.py (hunk 1 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py:208]: My agent says I replaced the unreachable false return with an explicit invariant check that raises AssertionError if the previous candidate ever appears in a later mutation region.

File: torch/_functorch/compile_utils.py (hunk 2 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py:455]: My agent says I now advance old_node_for_hash after successful same-region and cross-region substitutions, so future duplicates scan from the latest occurrence. I also added mutated-result and alias protections so this remains safe when duplicate outputs are later mutated.

File: torch/_functorch/compile_utils.py (hunk 3 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says _storage_refs_from_value now takes object instead of Any.

File: torch/_functorch/compile_utils.py (hunk 4 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says storage collection now uses tree_iter for this helper.

File: torch/_functorch/compile_utils.py (hunk 5 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says I tightened the try/except so it only covers untyped_storage(), then wraps the returned storage afterward.

File: torch/_functorch/compile_utils.py (hunk 3 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says _storage_refs_from_value now returns set[StorageWeakRef] | None instead of a tuple, with None representing unavailable storage metadata.

File: torch/_functorch/compile_utils.py (hunk 6 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says _collect_storage_refs_from_graph_values now uses tree_iter as suggested.

File: torch/_functorch/compile_utils.py (hunk 7 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says I moved the missing-argument sentinel to a module-level _MISSING.

File: torch/_functorch/compile_utils.py (hunk 8 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says _get_mutated_argument_values now returns tuple[Any, ...] | None.

File: torch/_functorch/compile_utils.py (hunk 9 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says I replaced the bounded for loop with while cur is not node, with the mutation-region invariant checked explicitly before the walk.

File: torch/_functorch/compile_utils.py (hunk 3 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says I added a comment explaining that StorageWeakRef is used here for storage identity/hashing without extending the underlying storage lifetime.

File: torch/_functorch/compile_utils.py (hunk 10 above)
Comment (jansel) [reply torch/_functorch/compile_utils.py]: My agent says I switched the alias check to not mutated_storage_refs.isdisjoint(storage_refs_to_protect).

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and resolved the CSE conflicts while preserving newer partition-key, stream/mempool, scalar-normalization, impurity, and functional-wrapper protections. The update also fails closed when tensor storage metadata is unavailable, adds a runtime-alias regression, and makes mutation-ancestor collection linear instead of per-mutation DFS. Local lint and targeted CPU/GPU tests pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
